### PR TITLE
improve search speed in side nav

### PR DESCRIPTION
**Pull request by XiaoxuHU**
This PR is for improving experience for using search box in Applens side nav

- Move filtering from each menu item to side-nav.component during searching
- Add a faster string comparison algorithm([KMP Algorithm](https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm)), to help improving search speed 

image

### CODE REVIEW

File: AngularApp/projects/applens/src/app/modules/dashboard/side-nav/side-nav.component.ts [hunk 1]
```diff
@@ -258,20 +261,69 @@ export class SideNavComponent implements OnInit {
...
+      const copiedItem = { ...item };
+      copiedItem.subItems = copiedSubItems;
+      res.push(copiedItem);
+    }
+    return res;
+  }
+
+  private checkMenuItemMatchesWithSearchTerm(item: CollapsibleMenuItem, searchValue: string) {
+    if (searchValue.length === 0) return true;
+    return StringUtilities.IndexOf(item.label.toLowerCase(), searchValue.toLowerCase()) >= 0 || StringUtilities.IndexOf(item.id.toLowerCase(), searchValue.toLowerCase()) >= 0 || (item.metadata && StringUtilities.IndexOf(item.metadata.toLowerCase(), searchValue.toLowerCase()) >= 0);
```
Comment (cindylovescoding) [inline AngularApp/projects/applens/src/app/modules/dashboard/side-nav/side-nav.component.ts:324]: We can probably remove the metadata comparison here

File: AngularApp/projects/applens/src/app/modules/dashboard/side-nav/side-nav.component.html [hunk 2]
```diff
@@ -22,8 +22,8 @@
           <ng-container *ngIf="analysisTypes.length > 0">
           <collapsible-menu-item [menuItem]="analysisTypes"></collapsible-menu-item>
         </ng-container>
-          <collapsible-menu-item *ngFor="let category of categories" [searchValue]="searchValue" [menuItem]="category"></collapsible-menu-item>
-          <i *ngIf="!categories || categories.length === 0" class="fa fa-spinner fa-spin fa-3x fa-fw loading-menu"></i>
+          <collapsible-menu-item *ngFor="let category of categories" [menuItem]="category" [searchValue]="searchValue"></collapsible-menu-item>
+          <i *ngIf="!categoriesCopy || categoriesCopy.length === 0" class="fa fa-spinner fa-spin fa-3x fa-fw loading-menu"></i>
```
Comment (cindylovescoding) [inline AngularApp/projects/applens/src/app/modules/dashboard/side-nav/side-nav.component.html:26]: Why do we want to use copy versions for categories and gists?

**Review by cindylovescoding — APPROVED** [review]

File: AngularApp/projects/applens/src/app/modules/dashboard/side-nav/side-nav.component.ts (hunk 1 above)
Comment (XiaoxuHU) [reply AngularApp/projects/applens/src/app/modules/dashboard/side-nav/side-nav.component.ts:324]: Sure, will do

File: AngularApp/projects/applens/src/app/modules/dashboard/side-nav/side-nav.component.html (hunk 2 above)
Comment (XiaoxuHU) [reply AngularApp/projects/applens/src/app/modules/dashboard/side-nav/side-nav.component.html:26]: Now, **categores** and **gists** are the item after filtering out by searching input
**categoriesCopy ** and **gistsCopy** are the all the items coming from backend response.
In here the i tag is the loading spinner when detector/gist is still under loading

File: AngularApp/projects/applens/src/app/modules/dashboard/side-nav/side-nav.component.ts [hunk 3]
```diff
@@ -258,20 +261,69 @@ export class SideNavComponent implements OnInit {
...
 
+  updateSearch(searchTerm: string) {
+    this.searchValue = searchTerm;
+    this.categories = this.updateMenuItems(this.categoriesCopy, searchTerm);
+    this.gists = this.updateMenuItems(this.gistsCopy, searchTerm);
+  }
+
+
+  //Only support filtering for two layer menu-item
+  private updateMenuItems(items: CollapsibleMenuItem[], searchValue: string): CollapsibleMenuItem[] {
```
Comment (XiaoxuHU) [inline AngularApp/projects/applens/src/app/modules/dashboard/side-nav/side-nav.component.ts:286]: I have another **recursive version** of this method for supporting updating items if we need to support more than three layers of menu items.

@cindylovescoding and @ShekharGupta1988, do you think we need that one or it is good enough for having current method for only supporting two layers of menu items ?

Comment (cindylovescoding) [reply AngularApp/projects/applens/src/app/modules/dashboard/side-nav/side-nav.component.ts:286]: I am not aware of scenarios that we need more than 3 layers. It should be fine

**Pull request merged**